File: examples/mouse.c
/**
 * Create a virtual mouse.
 * Compile me with: gcc mouse.c -o virtual_mouse -framework IOKit
 */

#include <IOKit/IOKitLib.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

unsigned char report_descriptor[] = {
    0x05, 0x01,                    // USAGE_PAGE (Generic Desktop)
    0x09, 0x02,                    // USAGE (Mouse)
    0xa1, 0x01,                    // COLLECTION (Application)
    0x09, 0x01,                    //   USAGE (Pointer)
    0xa1, 0x00,                    //   COLLECTION (Physical)
    0x05, 0x09,                    //     USAGE_PAGE (Button)
    0x19, 0x01,                    //     USAGE_MINIMUM (Button 1)
    0x29, 0x03,                    //     USAGE_MAXIMUM (Button 3)
    0x15, 0x00,                    //     LOGICAL_MINIMUM (0)
    0x25, 0x01,                    //     LOGICAL_MAXIMUM (1)
    0x95, 0x03,                    //     REPORT_COUNT (3)
    0x75, 0x01,                    //     REPORT_SIZE (1)
    0x81, 0x02,                    //     INPUT (Data,Var,Abs)
    0x95, 0x01,                    //     REPORT_COUNT (1)
    0x75, 0x05,                    //     REPORT_SIZE (5)
    0x81, 0x03,                    //     INPUT (Cnst,Var,Abs)
    0x05, 0x01,                    //     USAGE_PAGE (Generic Desktop)
    0x09, 0x30,                    //     USAGE (X)
    0x09, 0x31,                    //     USAGE (Y)
    0x15, 0x81,                    //     LOGICAL_MINIMUM (-127)
    0x25, 0x7f,                    //     LOGICAL_MAXIMUM (127)
    0x75, 0x08,                    //     REPORT_SIZE (8)
    0x95, 0x02,                    //     REPORT_COUNT (2)
    0x81, 0x06,                    //     INPUT (Data,Var,Rel)
    0xc0,                          //   END_COLLECTION
    0xc0                           // END_COLLECTION
};

struct mouse_report_t {
    uint8_t buttons;
    int8_t x;
    int8_t y;
};

#define SERVICE_NAME "it_unbit_foohid"

#define FOOHID_CREATE 0  // create selector
#define FOOHID_SEND 2  // send selector

#define DEVICE_NAME "Foohid Virtual Mouse"
#define DEVICE_SN "SN 123456"

int main() {
    io_iterator_t iterator;
    io_service_t service;
    io_connect_t connect;

    // Get a reference to the IOService
    kern_return_t ret = IOServiceGetMatchingServices(kIOMasterPortDefault, IOServiceMatching(SERVICE_NAME), &iterator);

    if (ret != KERN_SUCCESS) {
        printf("Unable to access IOService.\n");
        exit(1);
    }

    // Iterate till success
    int found = 0;
    while ((service = IOIteratorNext(iterator)) != IO_OBJECT_NULL) {
        ret = IOServiceOpen(service, mach_task_self(), 0, &connect);

        if (ret == KERN_SUCCESS) {
            found = 1;
            break;
        }

        IOObjectRelease(service);
    }
    IOObjectRelease(iterator);

    if (!found) {
        printf("Unable to open IOService.\n");
        exit(1);
    }

    // Fill up the input arguments.
    uint32_t input_count = 8;
    uint64_t input[input_count];
    input[0] = (uint64_t) strdup(DEVICE_NAME);  // device name
    input[1] = strlen((char *)input[0]);  // name length
    input[2] = (uint64_t) report_descriptor;  // report descriptor
    input[3] = sizeof(report_descriptor);  // report descriptor len
    input[4] = (uint64_t) strdup(DEVICE_SN);  // serial number
    input[5] = strlen((char *)input[4]);  // serial number len
    input[6] = (uint64_t) 2;  // vendor ID
    input[7] = (uint64_t) 3;  // device ID

    ret = IOConnectCallScalarMethod(connect, FOOHID_CREATE, input, input_count, NULL, 0);
    if (ret != KERN_SUCCESS) {
        printf("Unable to create HID device.\n");
        exit(1);
    }

    // Arguments to be passed through the HID message.
    struct mouse_report_t mouse;
    uint32_t send_count = 4;
    uint64_t send[send_count];
    send[0] = (uint64_t)input[0];  // device name
    send[1] = strlen((char *)input[0]);  // name length
    send[2] = (uint64_t) &mouse;  // mouse struct
    send[3] = sizeof(struct mouse_report_t);  // mouse struct len

    for(;;) {
        mouse.buttons = 0;
        mouse.x = rand();
        mouse.y = rand();

        ret = IOConnectCallScalarMethod(connect, FOOHID_SEND, send, send_count, NULL, 0);
        if (ret != KERN_SUCCESS) {
            printf("Unable to send message to HID device.\n");
        }

        sleep(1);  // sleep for a second
    }
}

File: examples/mouse.swift
/**
 * Create a virtual mouse.
 * To run:  swift mouse.swift
 * Stop with CTRL-C.
 * Install kernel extension from https://github.com/unbit/foohid/releases/latest first.
 */

import IOKit

let report_descriptor: [CUnsignedChar] = [
	0x05, 0x01,                    // USAGE_PAGE (Generic Desktop)
	0x09, 0x02,                    // USAGE (Mouse)
	0xa1, 0x01,                    // COLLECTION (Application)
	0x09, 0x01,                    //   USAGE (Pointer)
	0xa1, 0x00,                    //   COLLECTION (Physical)
	0x05, 0x09,                    //     USAGE_PAGE (Button)
	0x19, 0x01,                    //     USAGE_MINIMUM (Button 1)
	0x29, 0x03,                    //     USAGE_MAXIMUM (Button 3)
	0x15, 0x00,                    //     LOGICAL_MINIMUM (0)
	0x25, 0x01,                    //     LOGICAL_MAXIMUM (1)
	0x95, 0x03,                    //     REPORT_COUNT (3)
	0x75, 0x01,                    //     REPORT_SIZE (1)
	0x81, 0x02,                    //     INPUT (Data,Var,Abs)
	0x95, 0x01,                    //     REPORT_COUNT (1)
	0x75, 0x05,                    //     REPORT_SIZE (5)
	0x81, 0x03,                    //     INPUT (Cnst,Var,Abs)
	0x05, 0x01,                    //     USAGE_PAGE (Generic Desktop)
	0x09, 0x30,                    //     USAGE (X)
	0x09, 0x31,                    //     USAGE (Y)
	0x15, 0x81,                    //     LOGICAL_MINIMUM (-127)
	0x25, 0x7f,                    //     LOGICAL_MAXIMUM (127)
	0x75, 0x08,                    //     REPORT_SIZE (8)
	0x95, 0x02,                    //     REPORT_COUNT (2)
	0x81, 0x06,                    //     INPUT (Data,Var,Rel)
	0xc0,                          //   END_COLLECTION
	0xc0                           // END_COLLECTION
]

struct mouse_report_t {
	let buttons: UInt8
	let x: Int8
	let y: Int8
}

let SERVICE_NAME = "it_unbit_foohid"
let FOOHID_CREATE: UInt32 = 0 // create selector
let FOOHID_SEND:   UInt32 = 2 // send selector
let DEVICE_NAME =  "Foohid Virtual Mouse"
let DEVICE_SN =    "SN 123456"

enum VirtualInputError: Error, CustomStringConvertible {
	case notAccessIOService
	case notOpenIOService
	case notSendMessageToHID

	var description: String {
		switch self {
		case .notAccessIOService:
			return "Unable to access IOService."
		case .notOpenIOService:
			return "Unable to open IOService."
		case .notSendMessageToHID:
			return "Unable to send message to HID device."
		}
	}
}

func main() throws {
	var iterator: io_iterator_t = 0
	var connect: io_connect_t = 0

	// Get a reference to the IOService
	guard IOServiceGetMatchingServices(kIOMasterPortDefault, IOServiceMatching(SERVICE_NAME), &iterator) == KERN_SUCCESS
		else { throw VirtualInputError.notAccessIOService }
	defer { IOObjectRelease(iterator) }

	guard let service: io_object_t = (sequence(state: iterator) {
		let service = IOIteratorNext($0)
		return service != IO_OBJECT_NULL ? service : nil
		}.first {
			if IOServiceOpen($0, mach_task_self_, 0, &connect) == KERN_SUCCESS {
				return true
			}
			IOObjectRelease($0)
			return false
	}) else { throw VirtualInputError.notOpenIOService }
	defer { IOObjectRelease(service) }

	let DEVICE_NAME_POINTER = unsafeBitCast(strdup(DEVICE_NAME), to: UInt64.self)
	let DEVICE_NAME_LENGTH = UInt64(DEVICE_NAME.utf8.count)

	report_descriptor.withUnsafeBufferPointer { report_descriptor in
		// Fill up the input arguments.
		let input_count: UInt32 = 8
		var input = [UInt64](repeatElement(0, count: Int(input_count)))
		input[0] = DEVICE_NAME_POINTER	// device name
		input[1] = DEVICE_NAME_LENGTH // name length
		input[2] = unsafeBitCast(report_descriptor.baseAddress, to: UInt64.self) // report descriptor
		input[3] = UInt64(report_descriptor.count) // report descriptor len
		input[4] = unsafeBitCast(strdup(DEVICE_SN), to: UInt64.self) // serial number
		input[5] = UInt64(DEVICE_SN.utf8.count) // serial number len
		input[6] = 2 // vendor ID
		input[7] = 3 // device ID

		if IOConnectCallScalarMethod(connect, FOOHID_CREATE, input, input_count, nil, nil) != KERN_SUCCESS {
			print("Unable to create HID device. May be fine if created previously.")
		}
	}

	// Arguments to be passed through the HID message.
	let send_count: UInt32 = 4
	var send = [UInt64](repeatElement(0, count: Int(send_count)))
	send[0] = DEVICE_NAME_POINTER // device name
	send[1] = DEVICE_NAME_LENGTH // name length

	send[3] = UInt64(MemoryLayout<mouse_report_t>.stride) // mouse struct len

	while true {
		var mouse = mouse_report_t(buttons: 0, x: 20, y: 20)
		try withUnsafePointer(to: &mouse) { mouse in
			send[2] = UInt64(UInt(bitPattern: mouse)) // mouse struct
			guard IOConnectCallScalarMethod(connect, FOOHID_SEND, send, send_count, nil, nil) == KERN_SUCCESS
				else { throw VirtualInputError.notSendMessageToHID }
		}
		sleep(1)
	}
}

do {
	try main()
} catch {
	print(error)
	exit(1)
}

